File: backend/app/api/v1/routes/allocations.py
from __future__ import annotations

import uuid
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.authz.deps import get_authz_engine, require_permission
from app.authz.engine import AuthzEngine
from app.authz.enums import Action, FeatureKey
from app.core.pagination import Page
from app.database import get_db
from app.models.employee import Employee
from app.schemas.allocation import (
    AllocationCreate,
    AllocationPreview,
    AllocationUpdate,
    AllocationWriteResult,
)
from app.services import allocation_service

router = APIRouter(prefix="/allocations", tags=["allocations"])
FEATURE = FeatureKey.ALLOCATIONS


@router.get("", response_model=Page)
async def list_allocations(
    page: int = Query(1, ge=1),
    page_size: int = Query(50, ge=1, le=200),
    employee_id: uuid.UUID | None = Query(None),
    project_id: uuid.UUID | None = Query(None),
    active_on: date | None = Query(None),
    include_cancelled: bool = Query(False),
    overallocated_only: bool = Query(False),
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.VIEW)),
):
    items, total = await allocation_service.list_allocations(
        db,
        engine,
        current_employee,
        offset=(page - 1) * page_size,
        limit=page_size,
        employee_id=employee_id,
        project_id=project_id,
        active_on=active_on,
        include_cancelled=include_cancelled,
        overallocated_only=overallocated_only,
    )
    return Page(items=items, total=total, page=page, page_size=page_size)


@router.get("/options")
async def allocation_options(
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    actor: Employee = Depends(require_permission(FEATURE, Action.CREATE)),
):
    from sqlalchemy import select

    from app.authz.enums import RecordScope
    from app.authz.scope_filters import managed_project_ids_subquery
    from app.models.enums import ALLOCATABLE_PROJECT_STATUSES, EmploymentStatus
    from app.models.project import Project

    people = (
        await db.execute(
            select(Employee).where(Employee.employment_status == EmploymentStatus.ACTIVE).order_by(Employee.first_name)
        )
    ).scalars()
    query = select(Project).where(Project.approval_status == "approved", Project.status.in_(ALLOCATABLE_PROJECT_STATUSES))
    if await engine.get_scope(actor, FEATURE) != RecordScope.ALL:
        query = query.where(Project.id.in_(managed_project_ids_subquery(actor.id)))
    projects = (await db.execute(query.order_by(Project.name))).scalars()
    from app.services.reference_data_service import project_role_options

    return {
        "employees": [{"id": e.id, "label": e.full_name} for e in people],
        "projects": [{"id": p.id, "label": p.name} for p in projects],
        "project_roles": await project_role_options(db),
    }


@router.post("/preview")
async def preview_allocation(
    payload: AllocationPreview,
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    actor: Employee = Depends(require_permission(FEATURE, Action.CREATE)),
):
    return await allocation_service.preview_allocation(db, engine, actor, payload)


@router.get("/{allocation_id}")
async def get_allocation(
    allocation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.VIEW)),
):
    return await allocation_service.get_allocation(db, engine, current_employee, allocation_id)


@router.post("", response_model=AllocationWriteResult, status_code=201)
async def create_allocation(
    payload: AllocationCreate,
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.CREATE)),
):
    """
    Returns the saved allocation plus a capacity verdict. Exceeding 100% is a
    warning (`over_allocated: true`), not a rejection — see AllocationWriteResult.
    """
    return await allocation_service.create_allocation(db, engine, current_employee, payload)


@router.patch("/{allocation_id}", response_model=AllocationWriteResult)
async def update_allocation(
    allocation_id: uuid.UUID,
    payload: AllocationUpdate,
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.EDIT)),
):
    return await allocation_service.update_allocation(db, engine, current_employee, allocation_id, payload)


@router.post("/{allocation_id}/cancel")
async def cancel_allocation(
    allocation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.EDIT)),
):
    """
    Allocations are cancelled, never deleted, so the record of who was booked on
    what survives. Replaces the previous `DELETE /allocations/{id}`.
    """
    return await allocation_service.cancel_allocation(db, engine, current_employee, allocation_id)

File: frontend/src/api/allocations.ts
import { apiClient } from "./client";
import type { Allocation, AllocationWriteResult, Capacity, CapacityPeriod, Page } from "./types";

export interface AllocationListParams {
  page?: number;
  page_size?: number;
  employee_id?: string;
  project_id?: string;
  active_on?: string;
  include_cancelled?: boolean;
  overallocated_only?: boolean;
}

export interface AllocationPayload {
  employee_id: string;
  project_id: string;
  allocation_percent: number;
  role_on_project: string;
  start_date: string;
  end_date?: string | null;
  billable?: boolean;
  notes?: string | null;
  confirm_overallocation?: boolean;
  overallocation_reason?: string | null;
}

export interface AllocationPreview {
  over_allocated: boolean;
  total_allocation_percent: number;
  overallocated_periods: CapacityPeriod[];
  allocations: { allocation_id: string; project_name: string; manager_name: string; allocation_percent: number;
    start_date: string; end_date: string | null; status: string; over_allocated: boolean; total_allocation_percent: number }[];
}

export const allocationsApi = {
  options: () => apiClient.get<{
    employees: {id: string; label: string}[];
    projects: {id: string; label: string}[];
    project_roles: {id: string; value: string; label: string}[];
  }>("/allocations/options").then(r => r.data),
  preview: (payload: {employee_id: string; project_id: string; start_date: string; end_date: string | null;
    allocation_percent: number; exclude_allocation_id?: string}) =>
    apiClient.post<AllocationPreview>("/allocations/preview", payload).then(r => r.data),
  list: (params: AllocationListParams = {}) =>
    apiClient
      .get<Page<Allocation>>("/allocations", {
        params: { page: 1, page_size: 100, ...params },
      })
      .then((r) => r.data),
  get: (id: string) => apiClient.get<Allocation>(`/allocations/${id}`).then((r) => r.data),
  create: (payload: AllocationPayload) =>
    apiClient.post<AllocationWriteResult>("/allocations", payload).then((r) => r.data),
  update: (id: string, payload: Partial<AllocationPayload>) =>
    apiClient.patch<AllocationWriteResult>(`/allocations/${id}`, payload).then((r) => r.data),
  /** Allocations are cancelled, never deleted — history is preserved. */
  cancel: (id: string) =>
    apiClient.post<Allocation>(`/allocations/${id}/cancel`).then((r) => r.data),
  projectAllocations: (projectId: string) =>
    apiClient
      .get<Page<Allocation>>(`/projects/${projectId}/allocations`)
      .then((r) => r.data),
  employeeAllocations: (employeeId: string) =>
    apiClient
      .get<Page<Allocation>>(`/employees/${employeeId}/allocations`)
      .then((r) => r.data),
  capacity: (employeeId: string, on?: string) =>
    apiClient
      .get<Capacity>(`/employees/${employeeId}/capacity`, { params: on ? { on } : {} })
      .then((r) => r.data),
};
